TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/Node.php

```diff
@@ -761,43 +877,52 @@ public function hasProperty($propertyName)
+     * @param boolean $returnNodesAsIdentifiers If enabled, references to nodes are returned as node identifiers instead of NodeInterface instances
      * @return mixed value of the property
      * @api
      */
     public function getProperty($propertyName, $returnNodesAsIdentifiers = false)
     {
         $value = $this->nodeData->getProperty($propertyName);
-        if (!empty($value)) {
-            $nodeType = $this->getNodeType();
-            if ($nodeType->hasConfiguration('properties.' . $propertyName)) {
-                $expectedPropertyType = $nodeType->getPropertyType($propertyName);
-                switch ($expectedPropertyType) {
-                    case 'references' :
-                        if ($returnNodesAsIdentifiers === false) {
-                            $nodes = array();
-                            foreach ($value as $nodeIdentifier) {
-                                $node = $this->context->getNodeByIdentifier($nodeIdentifier);
-                                // $node can be NULL if the node is not visible according to the current content context:
-                                if ($node !== null) {
-                                    $nodes[] = $node;
-                                }
-                            }
-                            $value = $nodes;
-                        }
-                        break;
-                    case 'reference' :
-                        if ($returnNodesAsIdentifiers === false) {
-                            $value = $this->context->getNodeByIdentifier($value);
-                        }
-                        break;
-                    default:
-                        $value = $this->propertyMapper->convert($value, $expectedPropertyType);
-                        break;
-                }
-            }
+        if (empty($value)) {
+            return $value;
+        }
+
+        $nodeType = $this->getNodeType();
+        if (!$nodeType->hasConfiguration('properties.' . $propertyName)) {
+            return $value;
+        }
+
+        $expectedPropertyType = $nodeType->getPropertyType($propertyName);
+        if ($expectedPropertyType === 'references') {
+            return $this->resolvePropertyReferences($returnNodesAsIdentifiers, $value);
+        }
+
+        if ($expectedPropertyType === 'reference') {
+            return ($returnNodesAsIdentifiers ? $value : $this->context->getNodeByIdentifier($value));
+        }
+
+        return $this->propertyMapper->convert($value, $expectedPropertyType);
+    }
+
+    /**
+     * Maps the property value (an array of node identifiers) to the Node objects if needed.
+     *
+     * @param boolean $returnNodesAsIdentifiers
+     * @param array $value
+     * @return array
+     */
+    protected function resolvePropertyReferences($returnNodesAsIdentifiers, $value = [])
+    {
+        if ($returnNodesAsIdentifiers) {
+            return $value;
         }
```

Comment:
why not do that logic in ``getProperty`` like it's done with the ``reference`` type? seems odd sending it as a argument just to return the next argument if it's true..